lib/new_relic/ecto/telemetry.ex

```diff
@@ -0,0 +1,83 @@
+defmodule NewRelic.Ecto.Telemetry do
+  use GenServer
+
+  @handler_id :new_relic_ecto
+
+  def start_link(metrics: metrics) do
+    GenServer.start_link(__MODULE__, metrics: metrics)
+  end
+
+  def init(metrics: metrics) do
+    Process.flag(:trap_exit, true)
+
+    Enum.each(metrics, fn metric ->
+      :telemetry.attach(
+        @handler_id,
+        metric.event_name,
+        &__MODULE__.handle_event/4,
+        %{metric: metric}
+      )
+    end)
+
+    {:ok, %{metrics: metrics}}
+  end
+
+  def terminate(reason, _state) do
+    :telemetry.detach(@handler_id)
+  end
+
+  # TODO:
+  # * [x] Report DataStore metrics & aggregate
+  # * [ ] Report TT segments & DT spans
+  # * [x] Increment datastore_call_count, etc
+  # * [x] PR `repo` into ecto metadata
+
+  def handle_event(
```

Comment:
yeah quite possibly. I don't think I fully see how `telemetry_metrics` fits in yet :( I'm pondering moving all of this into the main `new_relic_agent` given how `telemetry` is structured (instead of a package-per-library model like this). Then I could have one single `NewRelic.Telemetry` module for all the event types. In that case it might be more relevant to use the `telemetry_metrics` package

I also don't really see why it's worth having the event be customizable in ecto - that just seems like it makes it kinda useless (hence the `_event` here). I'm relying exclusively on the metadata in this function, I don't even use the event name since it could be arbitrary

Reply:
I think moving it to a single package makes sense, since the whole point of telemetry project is to unify the process of gathering metrics across the ecosystem, no matter the library.

telemetry_metrics defines only what should be the result of aggregating specific events. For now it defines a counter, sum, last value (gauge) and distribution (histogram). The point here is that when a user passes a counter spec to the new relic reporter, they expect to see a metric which is a result of counting all the events with particular name. Or, when they provide last value, they want to see a metric which holds the value of the last event with that name.

I recommend reading the docs if you haven’t yet, and we can always chat on Elixir Slack 😄

Reply:
@binaryseed I have built a small StatsD reporter, you can check it out in this repo https://github.com/arkgil/telemetry_metrics_statsd 🙂

Reply:
Thanks, that's helpful
 
* NewRelic tries to provide a more "curated" experience, so we tend to decide the metrics we instrument internal to the agent. So it looks like some of what `telemetry_metrics` does is traditionally done inside our agent. I that's not a problem at all though, we can still manage our "curated" metrics and have a generic treatment for those that we don't know

* I see that your `handle_event` ignores the `_event`. I'm doing the same and I'm wondering if there is an intended purpose for that field that I'm missing..

* I was thinking about attaching each metric with it's own `handler_id` just like you do, so I think i'll adopt that approach

Reply:
@binaryseed yea, in this case, it made sense to ignore the event because in the handler config I have the information which metric should be updated. This is even more important when a single event is a source of two or more metrics. 

Let me know if you need any more help. I'm really glad that you are one of the early adopters of Telemetry 🙂